### PR TITLE
feat!: implement reconstruct for leopard ffe8

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

**Review by oblique — APPROVED** [review]
lgtm

File: src/lib.rs [hunk 1]
```diff
@@ -73,115 +100,253 @@ fn encode_inner(shards: &mut [&mut [u8]], data_shards: usize, shard_size: usize)
...
+    let full_passes = data_shards / m;
+    // if this is 'true', we would overflow the fft skew table which has the size of `MODULUS`
+    (full_passes + 1) * m + 1 > MODULUS as usize
+}
+
+/// Reconstructs the original shards from the provided slice.
+///
+/// The shards which are missing should be provided as empty `Vec`s.
+///
+/// Reconstruction can only happen if the amount of not missing data and parity shards
```
Comment (fl0rek) [inline src/lib.rs]:
```suggestion
/// Reconstruction can only happen if the amount of present data and parity shards
```
also, after reading this comment, word "shard" no [longer has any meaning to me ](https://en.wikipedia.org/wiki/Semantic_satiation) :P

File: src/lib.rs [hunk 2]
```diff
@@ -73,115 +100,253 @@ fn encode_inner(shards: &mut [&mut [u8]], data_shards: usize, shard_size: usize)
...
+    if m >= data_shards || last_count == 0 {
+        return false;
+    }
+
+    let full_passes = data_shards / m;
+    // if this is 'true', we would overflow the fft skew table which has the size of `MODULUS`
+    (full_passes + 1) * m + 1 > MODULUS as usize
+}
+
+/// Reconstructs the original shards from the provided slice.
```
Comment (fl0rek) [inline src/lib.rs]:
```suggestion
/// Reconstructs original shards from the provided slice.
```
I'm like 80% certain 🤔

**Review by fl0rek — APPROVED** [review]
Nice work!
wouldn't be myself without leaving a comment wording nit, though 😉

File: src/lib.rs [hunk 3]
```diff
@@ -53,12 +59,12 @@ pub type Result<T, E = LeopardError> = std::result::Result<T, E>;
...
 ///
 /// The first `data_shards` shards will be the treated as data shards
-/// and the rest as a parity shards.
+/// and the rest as parity shards.
 ///
 /// # Errors
 ///
 /// If too many shards provided or shards were of incorrect or different lengths.
-pub fn encode(shards: &mut [&mut [u8]], data_shards: usize) -> Result<()> {
+pub fn encode(shards: &mut [impl DerefMut<Target = [u8]>], data_shards: usize) -> Result<()> {
```
Comment (oblique) [inline src/lib.rs]: Much fewer types would implement `DerefMut` than `AsMut`.

**Review by oblique — APPROVED** [review]

**Pull request merged**